### PR TITLE
[4.0][Concept] Top menu option

### PR DESCRIPTION
**Pull request by ciar4n**
Pull Request for Issue # .

### Summary of Changes
**Disclaimer:** I'm not suggesting we should have this option, just that we can. Also, this is an option. The default is still in the sidebar. Requires further work if considered,

Creates a CSS class allowing the admin menu placed in the header similar to J3. This option is set via template style settings. It's just a CSS class so can be applied dynamically similar to the menu toggle.

### Testing Instructions
Apply PR and compile CSS (node build.js --compile-css). Navigate to Admin template settings and set **Menu Position** to **Top**

image

image

### CODE REVIEW

**Comment by wilsonge** [comment]
My personal opinion is that this is another one of those options that just clutters the interface. Just because we can doesn't mean we should. But if @coolcat-creations wishes to overrule me on that then that's ok

**Comment by 200MPHMEDIA** [comment]
For the NEW or NOVICE Joomla user, this menu addition would be continually helpful.  The left side menu is problematic anyway, because every time you click the "Control Panel" like, it generates the queries for Joomla, Extension and Over-ride updates - every single time, vs. once a day or only upon login - one instance.  So anything to avoid that, and this menu would help, would be welcome.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24063">issues.joomla.org/tracker/joomla-cms/24063</a>.</sub>

**Comment by brianteeman** [comment]
Please stop trolling. It has already been explained that it is not checking every time

**Comment by ciar4n** [comment]
I'm gonna close this. As stated, it is probably an option we can do without. I can re-open if required.

**Comment by 200MPHMEDIA** [comment]
It check it every time, or the boxes show that it is checking, every time I click on “Control Panel.”

 

So is it a display error with the messages?

**Comment by 200MPHMEDIA** [comment]
Want me to send you a log in to our back-end so you can see for yourself?

**Comment by infograf768** [comment]
@ciar4n 
In fact I do like your proposal. Much easier UI than this left side menu and does not cost much to implement. User should be able to choose.
**The main advantage is that icon AND text are always displayed, saving a click**
One aspect to solve though (among others) would be the dropdowns of the subs in RTL as they always display the Content Menu.

I guess that correcting creation of custom admin menus would be a similar task if menu is on top or side. It would be nice to solve that first imho.

@200MPHMEDIA 
The placement of the menu has no impact on the quickicons result.
Yes, it does check every time, but differently. After the first check the result should display almost immediately as it uses Ajax. 
Concerning extensions update cache see https://forum.joomla.org/viewtopic.php?t=952192#p3481181

**Comment by Joomv** [comment]
I love this as an option.  My biggest worry, being an ageing luddite is the loss of all the muscle memory over 15 years of top menu has given me (and yes I'm including Mambo in that timeline).  

I spend 12 hours a day on average in the back end of Joomla and the side menu is going to mess with my old brain big time.

**Comment by coolcat-creations** [comment]
The current template does not work with a horizontal menu. I agree that horizontal fits better to Joomla and will cause less confusion but that should have been considered back in 2016 not 2 months before beta. Additionally a lot of users asked in the last years for the horizontal menu and there was an absolute clear **NO** from production. I don't mind changing a concept, but not in this stage. 

Regarding Muscle Memory: Apple changed the direction of mouse scrolling. Lots of users were scared  but it took 3 days in average to learn the new direction. It will maybe take some time, but users will learn the new system. In this connection a nice on-boarding is important for Joomla. Maybe a former GSoC Project (Tour) can help.

**Comment by mbabker** [comment]
> In fact I do like your proposal. Much easier UI than this left side menu and does not cost much to implement. User should be able to choose.

The admin template should either be a sidebar or a top menu.  Not both.  Not an option to support both.  Either or.

Realistically, a CMS' admin interface does not need to be as flexible as its public facing counterpart.  There is debatably too much flexibility in the administrative interface in the first place and too many proposals to support too many other possible user level customizations.

**Comment by Joomv** [comment]
How about having 2 template views?
One for Horizontal one for Vertical and letting the users decide?
Ciaran's solution is a very simple thing and I do strongly believe it is the right way to go to give our users the option of having the admin the way they have had it for 14 years, or more including mambo, and giving them the option of trying something new.

Last time I checked we were not apple.  Forcing change is never a good thing.  Giving the option of change is always a good thing.

**Comment by mbabker** [comment]
Build two templates, one designed with a sidebar and one designed with the "classic" top menu.  I do not think it is in anyone's best interest to build an administrative template that supports both options.  The two approaches to a navigation system call for different organizational structures.  A sidebar system enables you to "promote" more items to the root of your menu tree, a top menu forces you to be selective on how many root items exist (or even the labels used, as has been an issue in maintaining both Isis and the `joomla.org` top menus otherwise you end up with text overflow or unintentional multi-line menus in non-English locales).

Neither solution is "right", I'll go so far as to say neither solution is "better", but I honestly believe an administrative interface supporting both options is just asking for trouble.

**Comment by brianteeman** [comment]
Old people always find change hard at first but it doesn't take long for people to forget all about the old way.

Remember when the toolbar switched from RTL to LTR

Or when we went from two to one text editor window.

So many things change

**Comment by coolcat-creations** [comment]
> How about having 2 template views?
> One for Horizontal one for Vertical and letting the users decide?
> Ciaran's solution is a very simple thing and I do strongly believe it is the right way to go to give our users the option of having the admin the way they have had it for 14 years, or more including mambo, and giving them the option of trying something new.
> 
> Last time I checked we were not apple. Forcing change is never a good thing. Giving the option of change is always a good thing.

If someone has time a Design and PR for 4.1 it can be done for 2 complete different views/templates. But mind the pitfalls that will come if it comes to make both views work with the structure of the system in every case and update and maintaining two views. I just say it's not possible at this point and timeline and ressources.

**Comment by ciar4n** [comment]
> If someone has time a Design and PR for 4.1 it can be done for 2 complete different views/templates.

With the template in its current form, there is no need for this. As demonstrated here, the top menu can be implemented by simply adding a class to the body tag. How that tag is added can be done via a setting (this PR) or by JS similar to the current menu toggle button.

**Comment by coolcat-creations** [comment]
TBH it does not fit into the template - You would need to do something with the topbar - moving the icons and version to bottom and put the Menu at the very top. I would rather not provide a "halfgood" solution just because. Thats just my opinion.

**Comment by ciar4n** [comment]
>  You would need to do something with the topbar - moving the icons and version to bottom and put the Menu at the very top.

All this can be equally done with just the addition of a class to the body. But as I said in opening this PR, I'm not suggesting this should be done, just that it can.

**Comment by infograf768** [comment]
> Old people always find change hard at first

That is ridiculous and insulting. Please stop that kind of comment.

The real point is **"How easy is it to move in the admin between different managers"**
Taking off the sidebar even adds more clicking if for the same component.

Basically, if the menu is placed on top AND we keep showing items with Text and their submenus (This is only partly implemented here with this patch as the Text is hidden once you display the manager [would need some modifications in the js] and submenus are not displayed by hovering the menus), we avoid multiple clicks.

Example with present menu in 4.0 for the same component:
sidemenu

As it is now, for the same component with different managers (without the sidebar), we need at least 3 clicks.

If we want to display another component submenu we need at least 4 clicks. 

sidemenu2

It is much simpler in 3.x, old or not...

**Comment by coolcat-creations** [comment]
The current behavior of the menu needs indeed improvement. Should we maybe focus on that? And talk about horizontal for 4.1?

**Comment by infograf768** [comment]
No problem for me, as long as there is a real improvement in the left side menu.

**Comment by ciar4n** [comment]
> The current behavior of the menu needs indeed improvement. Should we maybe focus on that? And talk about horizontal for 4.1?

I think taking away the horizontal menu in 4.0 and then adding it back in 4.1 would be a bad idea. It will just show us back-tracking and not knowing what we want.

**Comment by coolcat-creations** [comment]
If - whoever has time wants to make the horizontal menu fit:
- [ ] the pagetitle has to be moved out of the topbar
- [ ] the Icons have to be moved elsewhere. 
- [ ] A new bottombar with the Version and the additional plugins has to appear. 
- [ ] Implement the Dashboards in Menu
- [ ] Implement the Add icons in Menu

But just for the case that the menu is horizontal, for vertical it should stay like its now. Both has to work in mobile too. A decision would need to be done which layout is used for marketing and docs.

Still convinced that we should use the time to make the vertical work.

**Comment by infograf768** [comment]
> It will just show us back-tracking and not knowing what we want.

If it is a parameter I see no issue. Just a choice.
If the original idea was wrong, it is still time to correct.
There are many other aspects of 4.0 which needs going back (or rather correcting) taking into account what was working correctly and is no more: joomla installation is a good example as it is very much broken. Custom admin menus is another aspect. Etc.

**Comment by mbabker** [comment]
> If it is a parameter I see no issue. Just a choice.

I'm going to tell you all the same thing @brianteeman has been yelling at me about for the last decade.  Stop thinking about this from purely the code perspective.  @ciar4n demonstrated the code is rather lightweight if the project is really serious about pursuing support of this (IMO fatally flawed) option.  This PR is more a perfect case of "just because you can doesn't mean you should" than a lot of things I have seen (or even suggested myself) thrown out there.  If you all are hell bent on having this option because "muscle memory is good and change is bad", are you all seriously willing to commit to supporting a menu system that is by default optimized for both a horizontal AND vertical variation?  Which essentially means the horizontal structure is going to dictate how the vertical sidebar is built, because if you're going to support both then in essence you have to optimize for the least common denominator, which is a horizontal top menu which has less available space than a vertical sidebar.  At which point you should just abandon the sidebar because you lose all benefits of shipping it.

So again, pick one and stick with it.  I know a lot of you like to ignore me as shooting down ideas, but this is one of those times where I am telling you that having this type of option is just going to make people's lives miserable.

**Comment by infograf768** [comment]
I understand what you say, but please stop the “muscle memory” or “old people” stuff...
Read my lips again: 
My comments were mostly related to the poor UI of the present admin menus forcing people to multiply clicks.
It did look simpler to me to use horizontal menus instead of what we have to reach this goal.

**Comment by 200MPHMEDIA** [comment]
Chiming in with an agreement:  left menu requires admin to perform too many clicks and back & forth when adding a lot of content.  Could have been more streamlined.  Understand the left menu is better for mobile, but how many people actually administer a Joomla! website via a mobile phone vs. laptops or desktop?

**Comment by mbabker** [comment]
> My comments were mostly related to the poor UI of the present admin menus forcing people to multiply clicks.

That is because of the heavy insistence on removing the secondary level sidebars (AKA component sidebar) in favor of everything using a single menu element, which has its pros and cons.  Switching to a horizontal menu does not change that workflow or improve the number of clicks needed to navigate to another section; rather the secondary level sidebar would need to be re-introduced, which from everything I have gathered from the designers in this repository is not an option (and if anything the design proposals seem to indicate the number of clicks it'll take to move about the admin will only increase as the proposal is actually implemented).

> Chiming in with an agreement: left menu requires admin to perform too many clicks and back & forth when adding a lot of content.

Again, this is an implementation/workflow flaw, not a "sidebar navigation for a user interface is broken" flaw.  If you do an analysis of administrative interfaces (both CMS specific and general purpose templates like [AdminLTE](https://adminlte.io/)), how many implement a sidebar navigation as the primary navigation element?  How many implement a top menu?  How many have support for both (the option proposal from this PR)?  Has anyone looked into the reasoning these structures are used and/or preferred, or is the argument here really boiling down to "Joomla has always had a top menu therefore it should not change"?

I'll be honest, I don't give a damn whether a top menu or sidebar is used.  I do give a damn if you all are crazy enough to have an option to support both.  Because I genuinely believe that is setting up your support groups for massive failure for the lifetime of this template.

**Comment by brianteeman** [comment]
Surely the time to make these objections was when the template was first presented

**Comment by wilsonge** [comment]
I'm locking this. If people feel really strongly about this and have fresh opinions then come message me or email me. For those not actively involved in Joomla my contact details are available on my volunteers portal page https://volunteers.joomla.org/joomlers/160-george-wilson
